File: app/src/main/java/ptv/example/zoulinheng/androidutils/utils/helpers/PermissionHelper.java
package ptv.example.zoulinheng.androidutils.utils.helpers;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

public class PermissionHelper {
    /**
     * 获取请求权限中需要授权的权限（禁止 和 询问）
     *
     * @param context     上下文
     * @param permissions 查询目标集合
     * @return 需要授权的权限集合
     */
    public static String[] getDeniedPermissions(final Context context, final String[] permissions) {
        List<String> deniedPermissions = new ArrayList<>();
        for (String permission : permissions) {
            if (ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_DENIED) {
                deniedPermissions.add(permission);
            }
        }
        return deniedPermissions.toArray(new String[deniedPermissions.size()]);
    }

    /**
     * 是否包含彻底拒绝了的权限
     */
    public static boolean hasAlwaysDeniedPermission(final Context context, final String... deniedPermissions) {
        for (String deniedPermission : deniedPermissions) {
            if (!isHasPermission(context, deniedPermission)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 判断该权限是否已经被授予
     *
     * @param permission 待判断的权限
     * @return true 已经授予该权限 ，false未授予该权限
     */
    private static boolean isHasPermission(Context context, String permission) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M)
            return true;
        return ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED;
    }

    /**
     * 获取权限描述数组
     *
     * @param permissions 权限数组
     * @return
     */
    public static String[] getPermissionNames(String... permissions) {
        List<String> perNameList = new ArrayList<>();
        for (String permission : permissions) {
            String permissionName = getPermissionName(permission);
            if (!TextUtils.isEmpty(permissionName)) {
                perNameList.add(permissionName);
            }
        }
        return perNameList.toArray(new String[perNameList.size()]);
    }

    /**
     * 获取权限描述
     *
     * @param permission 权限
     * @return 描述
     */
    private static String getPermissionName(String permission) {
        switch (permission) {
            case Manifest.permission.CALL_PHONE:
                return "电话权限";
            case Manifest.permission.CAMERA:
                return "相机权限";
            case Manifest.permission.ACCESS_FINE_LOCATION:
                return "位置权限";
            case Manifest.permission.READ_EXTERNAL_STORAGE:
                return "存储权限";
            case Manifest.permission.READ_PHONE_STATE:
                return "手机状态";
            default:
                return "";
        }
    }

    /**
     * 获取已禁止再次询问的权限
     *
     * @param context     上下文
     * @param permissions 权限集合源
     * @return
     */
    public static String[] getAllAlwaysDeniedPermission(Context context, String... permissions) {
        //需要提示的权限（询问）
        String[] shouldShowPermissions = getShouldShowRequestPermissionArray(context, permissions);
        //需要授权的权限（禁止和询问）
        String[] deniedPermissions = getDeniedPermissions(context, permissions);

        List<String> pers = new ArrayList<>();

        for (String s : deniedPermissions) {
            if (!isExist(shouldShowPermissions, s)) {
                pers.add(s);
            }
        }
        return pers.toArray(new String[pers.size()]);
    }

    /**
     * 获取需要权限需要说明提示的权限集合
     *
     * @param context           上下文
     * @param deniedPermissions 查询目标集合
     * @return
     */
    public static String[] getShouldShowRequestPermissionArray(Context context, String... deniedPermissions) {
        List<String> pers = new ArrayList<>();
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M)
            return pers.toArray(new String[pers.size()]);
        String[] deniedPers = getDeniedPermissions(context, deniedPermissions);
        for (String permission : deniedPers) {
            if (ActivityCompat.shouldShowRequestPermissionRationale((Activity) context, permission)) {
                pers.add(permission);
            }
        }
        return pers.toArray(new String[pers.size()]);
    }

    private static boolean isExist(String[] arr, String targetValue) {
        for (String s : arr) {
            if (s.equals(targetValue))
                return true;
        }
        return false;
    }
}

File: app/src/main/java/ptv/example/zoulinheng/androidutils/utils/permissions/PermissionHelper.kt
package ptv.example.zoulinheng.androidutils.utils.permissions

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.tbruyelle.rxpermissions2.RxPermissions
import ptv.example.zoulinheng.androidutils.utils.Utils
import java.util.ArrayList

/**
 * Created by zoulinheng on 2020/2/24.
 * desc: 权限操作类
 */
class PermissionHelper {
    companion object {

        /**
         * 申请权限
         * @param activity FragmentActivity
         * @param callBack 回调
         * @param permissions 需要申请的权限
         */
        @SuppressLint("CheckResult")
        fun permissions(activity: FragmentActivity, callBack: CallBack, vararg permissions: String) {
            val rxPermissions = RxPermissions(activity)
            rxPermissions.requestEachCombined(*permissions)
                    ?.subscribe { permission ->
                        when {
                            permission.granted -> {//全部权限均已获取
                                val grantedList = getDeniedPermissions(*permission.name.split(",").toTypedArray())
                                callBack.onGranted(grantedList)
                            }
                            permission.shouldShowRequestPermissionRationale -> {//有需要询问的权限
                                val shouldShowList = getNeedToAskPermissionArray(activity, *permission.name.split(",").toTypedArray())
                                callBack.shouldShowRequest(shouldShowList)
                            }
                            else -> {//此处对已不再询问的权限做操作
                                val noAgainAskList = getNoAskAgainPermissionArray(activity, *permission.name.split(",").toTypedArray())
                                callBack.noAskAgain(noAgainAskList)
//                                toSettingAct(activity)
                            }
                        }
                    }
        }

        private fun toSettingAct(activity: Activity) {
            activity.startActivity(PermissionsPageManager.getIntent(activity))
        }

        /**
         * 获取请求权限中未获取的权限集合（禁止 和 询问）
         *
         * @param permissions 查询目标集合
         * @return 需要授权的权限集合
         */
        private fun getDeniedPermissions(vararg permissions: String): Array<String> {
            val deniedPermissions: MutableList<String> = ArrayList()
            for (permission in permissions) {
                if (isGranted(permission)) {
                    deniedPermissions.add(permission)
                }
            }
            return deniedPermissions.toTypedArray()
        }

        /**
         * 获取需要询问的权限集合（询问）
         *
         * @param context     上下文
         * @param permissions 查询目标集合
         * @return 需要询问的权限集合
         */
        fun getNeedToAskPermissionArray(context: Context, vararg permissions: String): Array<String> {
            val pers: MutableList<String> = ArrayList()
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return pers.toTypedArray()
            val deniedPers = getDeniedPermissions(*permissions)
            for (permission in deniedPers) {
                if (shouldShowRequestPermissionRationale(context as Activity, permission)) {
                    pers.add(permission)
                }
            }
            return pers.toTypedArray()
        }

        /**
         * 获取被禁止的权限集合（询问）
         *
         * @param context     上下文
         * @param permissions 查询目标集合
         * @return 被禁止的权限集合
         */
        fun getNoAskAgainPermissionArray(context: Context, vararg permissions: String): Array<String> { //未获取的权限集合（禁止和询问）
            val unDeniedPermissions = getDeniedPermissions(*permissions)
            //需要询问的权限集合（询问）
            val shouldShowPermissions = getNeedToAskPermissionArray(context, *permissions)
            val pers: MutableList<String> = ArrayList()
            for (s in unDeniedPermissions) {
                if (!shouldShowPermissions.contains(s)) {
                    pers.add(s)
                }
            }
            return pers.toTypedArray()
        }

        private fun isGranted(permission: String): Boolean {
            return (Build.VERSION.SDK_INT < Build.VERSION_CODES.M
                    || PackageManager.PERMISSION_GRANTED == ContextCompat.checkSelfPermission(Utils.getApp(), permission))
        }

        private fun shouldShowRequestPermissionRationale(activity: Activity, permission: String): Boolean {
            return ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)
        }
    }

    interface CallBack {
        fun onGranted(permissions: Array<String>)
        fun shouldShowRequest(permissions: Array<String>)
        fun noAskAgain(permissions: Array<String>)
    }
}